partiql-lang/src/main/resources/org/partiql/schemas/graph.isl

```diff
@@ -0,0 +1,71 @@
+$ion_schema_2_0
+
+/** This ISL schema is a specification of an Ion-based file format for defining graphs externally,
+    in order to be brought into PartiQL data model for querying with the PartQL implementation of GPML.
+    This is useful for general reasons of accessing external data, but is particularly necessary
+    for graphs, since PartiQL does not yet define concrete syntax for them.
+
+    The design of the schema closely follows the graph data model description in
+    PartiQL [RFC-0025](https://github.com/partiql/partiql-docs/blob/main/RFCs/0025-graph-data-model.md),
+    with the exception that the data model in this format is not recursive: the payloads at nodes and edges
+    are general Ion.  That is, it is not specified how a fragment of a payload can be itself recognized as a graph.
+
+    A file contains a graph if it contains a single Ion struct of ISL type Graph, as defined here.
+
+    Annotations specified in this schema (graph, node, edge, dir, undir) are for mnemonic purposes only --
+    graph authors are not required to use them and processors should not attribute semantic meaning to their
+    presence or absence.
+*/
+
+type::{
+    name: graph,  // ExternalGraphValidationTests breaks when this is named Graph ... no idea
+    annotations: closed::[graph],
+    type: struct,
+    fields: closed::{
+        nodes: { type: list, element: Node, occurs: required },
+        edges: { type: list, element: Edge, occurs: required }
+    }
+}
+
+type::{
+    name: Node,
+    annotations: closed::[node],
+    type: struct,
+    fields: closed::{
+        id: { type: NodeId, occurs: required },
+        labels: { type: list, element: string, occurs: optional },
+        payload: { type: NodePayload, occurs: optional }
+    }
+}
+
+type::{
+    name: Edge,
+    annotations: closed::[edge],
+    type: struct,
+    fields: closed::{
+        id: { type: EdgeId, occurs: required },
+        labels: { type: list, element: string},
+        ends: { one_of: [DirectedPair, UndirectedPair], occurs: required },
+        payload: { type: EdgePayload, occurs: optional }
+    }
+}
+
+type::{
+    name: DirectedPair,
+    type: list,
+    ordered_elements: [ NodeId, NodeId ],
+    annotations: closed::[dir]
+}
+
+type::{
+    name: UndirectedPair,
+    type: sexp,
```

Comment:
Because annotations have no semantic meaning, does that mean the distinction between directed and undirected pairs is simply the bracketing?
```ion
[ a, b ]  // directed
( a, b )  // undirected
```

This is a bit subtle and since both ion lists and s-expressions are ordered there isn't necessarily a relationship between directed and undirected to the orderedness of the structure representing the pair. Unfortunately Ion doesn't have an unordered collection, but I think using s-expression operators you could improve the visual clues

```ion
( a -> b )  // directed, also explicitly shows direction
( a -- b )  // undirected
```

This is valid and Ion and is consistent with [Graphviz DOT](https://graphviz.org/doc/info/lang.html)

> An edgeop is -> in directed graphs and -- in undirected graphs.

Reply:
I like this idea!  

The `[ ]` vs `( )` was indeed too clever.  All I liked it for was avoiding a required heavy-weight discriminator.

Reply:
Done.  

I went for an extra syntactic frill of allowing differing length: `<-` and `<--`, etc.